hub.jupytearth.org-image/environment.yml

```
@@ -0,0 +1,59 @@
+channels:
+  - conda-forge
+dependencies:
+  - altair
+  - bqplot
+  - python-kaleido
+  - ipycanvas
+  - ipympl
+  - cython
+  - jupyter_bokeh
+  - jupyterlab-geojson
+  - sympy
+  - proj-data
+  # Dev tools
+  - black
+  - flake8
+  - pep8
+  - pyflakes
+  - pylint
+  - pytest
+  - pytest-cov
+  # jupyter-book is disabled as its providing upper bounds on
+  # its dependencies, and that has led to ipywidgets 8 was
+  # held back.
+  # https://github.com/executablebooks/jupyter-book/issues/1898.
+  # - jupyter-book
+  - jupytext
+  - numpydoc
+  - sphinx
+  # Data visualization
+  - plotly
+  - ipydatagrid
+  - ipyparallel
+  - pyhdf
+  - vaex
+  - mhealpy
+  - pytables
+  - xlrd
+  - jupyter-repo2docker
+  # UI changes
+  - jupyter-vscode-proxy
+  - jupypterlab-link-share
+  - jupyterlab-system-monitor
+  - jupyterlab-favorites
+  - nbdime
+  - retrolab
+  - ipydrawio
+  - google-cloud-sdk
+  - qgis
+  - syncthing
+  - websockify
+  - firefox
+  - pip
+  - pip:
+      - jupyter-remote-desktop-proxy==1.0.0
```

I'm very opinionated that these kinds of pins to patch version, or even minor version for something expecting to follow semver, shouldn't be done unless its by automation creating a pip freeze or equivalent output where all dependencies gets pinned to the same level.

```suggestion
      - jupyter-remote-desktop-proxy
```

---

I think some pins are better than none, but i don’t want to block this PR - ultimately proper freezing is the right call. So i will just accept this suggestion.

---

@consideRatio I wrote up what I meant here in a little bit more detail. No action needed, but I hope this makes you feel a bit more heard and not dismissed! 

Three broad kinds of version specification:

1. No pinning - every rebuild *will* grab new versions of a *lot* of things. We don't know what things are coming, so surprises await! Example: This repo.
2. First-level pinning only - every rebuild *will* grab new versions of *some* things, primarily things that are transitive dependencies. We don't *know* what new things are coming at all, so surprises await! However, *often*, because numpy / scipy / similar packages that are the cause of most breakage *are* pinned, the breakages are fewer than they would otherwise be. But when it does happen, you'll need to compare the output of `pip freeze` to go figure out what happens. Also depending on what you are installing and how, it is possible that the final installed version would actually *not* be the version you have specified manually! Example: [All the UC Berkeley Images](https://github.com/berkeley-dsep-infra/datahub/), the [2i2c default image](https://github.com/2i2c-org/2i2c-hubs-image/)
3. Minimal versioning specified (usually upper bounds when there are known issues) manually, automatic lock file generated to be installed from. This approach is similar to (1), so versions are often not specified. **However**, in CI, a bot will use a tool like `conda-lock` or `pip-tools` to explicitly resolve the versions of packages that will be installed, and commit them to the repository. This means that each build we will know *exactly* what will change. And if we don't want a particular change to happen, we can put an upper bound or something on it as a choice. Example: Pangeo images

I think (3) is clearly superior, but it requires the following bits of engineering work to be done:

1. A CI/CD bot that does the locking mechanism and commits this to CI. This is important because asking people to run `conda-lock` or similar locally doesn't quite work well.
2. If we are inheriting from a base image, *that* image also has an existing conda environment usually. And so the locking will need to take into account the existing versions of packages there as well. This will need engineering work to be done.

Based on personal experience, and while the engineering pieces required for (3) are not available, I prefer (2) over (1) because (2) has *chances* of version upgrades while (1) *guarantees* version upgrades. If there is breakage due to transitive dependencies, `diff`ing `pip freeze` or similar usually helps spot the culprit. This kinda `diff`ing is *required* with (3) as well - although it has the advantage of being done *before* merging rather than after :)

There is a *fourth* method, which is that *only* the lock file equivalent is committed to the repository, and sometimes this is even hand written! Sometimes folks make a local conda env, and use `conda export`. This I would recommend against ever doing for the following reasons:

1. This basically sidesteps the dependency solver present in pip / conda, so *any* upgrades to any versions of any package are going to be super brittle! Anyone trying to edit this list of packages later on will have to do the work of `conda-lock` or `pip-tools` manually. Two packages might have a common transitive dependency, and they might have compatible version requirements, but not the latest. For the dependency solver, this is nbd. For a human, this is quite a lot of busywork. 
2. There is no marker for what packages are being *intentionally* installed vs installed as transitive dependencies. This makes packages really difficult to remove!

Sometimes in the absense of tooling for (3), people do this (4). (4) is strictly the worst of all worlds, and let's never do that. Example, old data100 image.

---

Thank you for your thoughtful response @yuvipanda!!

> 3. Minimal versioning specified (usually upper bounds when there are known issues) manually, automatic lock file generated to be installed from.

> I think (3) is clearly superior, but it requires the following bits of engineering work to be done:

We are in full agreement here!

I think I appreciate option 1 and 3 quite a lot because how well defined they are in comparison to option 2. Option 2 could for example involve precautionary pinning to either major, minor, or patch, and option 2 could be partially or fully adopted by pinning some but not all specified dependencies. So when speaking about option 2, those details matter also.

I think overall I prefer option 1 and 3 over option 2, and I'm currently far more favorable to an option 2 variant that only pins to the major like version (`1.*`, `0.1.*`) consistently or partially than an option 2 that precautionary pins to patch versions. If option 2 with patch version pinning starts to seem important for stability, I think option 3 should be adopted at that point.

I'd appreciate if we have a sync topic discussion about it where we collaborate on exploring the pro's and con's and trying to weigh them if you think option 2 with patch version pinning is relevant to persue instead of option 3.